python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -210,66 +219,74 @@ def _update(self):
-            if ((should_keep_worker_of_node_type
-                 or not nodes_allowed_to_terminate.get(node_id, True))
+            should_keep_or_terminate = self._keep_worker_of_node_type(
+                node_id, node_type_counts)
+            if ((should_keep_or_terminate == KeepOrTerminate.keep
+                 or node_id in nodes_not_allowed_to_terminate)
                     and self.launch_config_ok(node_id)):
+                keep_node(node_id)
+                continue
+            if should_keep_or_terminate == KeepOrTerminate.terminate:
+                schedule_node_termination(node_id, "max_workers_per_type")
                 continue
 
             node_ip = self.provider.internal_ip(node_id)
             if node_ip in last_used and last_used[node_ip] < horizon:
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating idle node.".format(node_id))
-                self.event_summarizer.add(
-                    "Removing {} nodes of type " + self._get_node_type(node_id)
-                    + " (idle).",
-                    quantity=1,
-                    aggregate=operator.add)
-                nodes_to_terminate.append(node_id)
+                schedule_node_termination(node_id, "idle")
             elif not self.launch_config_ok(node_id):
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating outdated node.".format(node_id))
-                self.event_summarizer.add(
-                    "Removing {} nodes of type " + self._get_node_type(node_id)
-                    + " (outdated).",
-                    quantity=1,
-                    aggregate=operator.add)
-                nodes_to_terminate.append(node_id)
-
-        if nodes_to_terminate:
-            self._terminate_nodes_and_cleanup(nodes_to_terminate)
-            nodes = self.workers()
+                schedule_node_termination(node_id, "outdated")
+            else:
+                keep_node(node_id)
+                nodes_we_could_terminate.append(node_id)
 
         # Terminate nodes if there are too many
-        nodes_to_terminate = []
-        while (len(nodes) -
-               len(nodes_to_terminate)) > self.config["max_workers"] and nodes:
-            to_terminate = nodes.pop()
-            logger.info("StandardAutoscaler: "
-                        "{}: Terminating unneeded node.".format(to_terminate))
-            self.event_summarizer.add(
-                "Removing {} nodes of type " +
-                self._get_node_type(to_terminate) + " (max workers).",
-                quantity=1,
-                aggregate=operator.add)
-            nodes_to_terminate.append(to_terminate)
+        num_extra_nodes_to_terminate = (
+            len(nodes) - len(nodes_to_terminate) - self.config["max_workers"])
+        num_extra_nodes_to_terminate = min(num_extra_nodes_to_terminate,
+                                           len(nodes_we_could_terminate))
```

Comment:
having a hard time here. Shouldn't num_extra_nodes_to_terminate < nodes_we_could_terminate? 
if not then there is a bug because max_workers should take precedence over all the other constraints.

Reply:
We have the following:
len(nodes): number of nodes at the beginning of _update
len(nodes_to_terminate): number of nodes we are terminating for sure.
The difference len(nodes) - len(nodes_to_terminate) is the number of live nodes if we do not do anything to enforce max_workers.
nodes_we_could_terminate: nodes that got marked with decide_later, ordered by last used time (oldest first).
If we have too many nodes (len(nodes) - len(nodes_to_terminate) > max_workers), we want to terminate some nodes from the tail of nodes_we_could_terminate.

if the configs (resource requests, min_workers) are consistent, we should have num_extra_nodes_to_terminate <= nodes_we_could_terminate, but I would rather not rely on that here.

Reply:
Should we add a warning/error message if this was inconsistent ? 
maybe add a comment saying that this is a sanity check ?

Reply:
Adding a comment here.
I think we should have a consistency check when loading configs.